### DISCUSSION TITLE
Is it possible to process an image locally on a webpage using JS instead of uploading it to a web server?

### DISCUSSION THREAD

**Discussion by SamJones04** [community/community#132090] [Programming Help] [answered]
So basically the issue I'm having is that I'm trying to make a **PDF417 Barcode reader** in a website, the site is hosted on GitHub Pages so I have no infrastructure of my own and I'm unable to upload images.

I attempted to use zxing reader and Dynamsofts reader (using the live video stream from the camera on the phone) but both options don't recognise my example barcodes at all. Is there any actually functional PDF417 readers in JavaScript that support live camera streams?

**Comment by RachidZahrani** [comment]
For implementing a PDF417 barcode reader on a website hosted on GitHub Pages, you can use JavaScript libraries that support live camera streams.

**Comment by tienanh109** [comment, marked as answer]
Yes, it is possible to process an image locally on a webpage using JavaScript. You can use JavaScript libraries that allow for image processing directly in the browser without the need to upload the image to a server. For reading PDF417 barcodes, you can use libraries such as `quaggaJS` or `pdf417-js`.

Here’s a basic example using `pdf417-js` to read a barcode from a live video stream:

1. **Include the `pdf417-js` library**:
   You can include it in your HTML file directly from a CDN.

   ```html
   <script src="https://cdn.jsdelivr.net/npm/pdf417-js@0.0.1/build/pdf417.min.js"></script>
   ```

2. **Set up the HTML**:
   You need a video element to display the live stream from the camera.

   ```html
   <video id="video" width="600" height="400" autoplay></video>
   <canvas id="canvas" style="display:none;"></canvas>
   ```

3. **JavaScript to access the camera and process the video stream**:
   This script accesses the camera, captures frames from the video stream, and uses the `pdf417-js` library to decode the barcode.

   ```html
   <script>
     const video = document.getElementById('video');
     const canvas = document.getElementById('canvas');
     const context = canvas.getContext('2d');

     // Access the device camera and stream to video element
     navigator.mediaDevices.getUserMedia({ video: true })
       .then(stream => {
         video.srcObject = stream;
         video.play();
       })
       .catch(err => console.error('Error accessing the camera', err));

     // Process the video frame and decode the barcode
     video.addEventListener('play', () => {
       const scanBarcode = () => {
         if (video.paused || video.ended) return;
         canvas.width = video.videoWidth;
         canvas.height = video.videoHeight;
         context.drawImage(video, 0, 0, canvas.width, canvas.height);
         const imageData = context.getImageData(0, 0, canvas.width, canvas.height);
         try {
           const decoded = PDF417Decoder.decode(imageData.data, canvas.width, canvas.height);
           console.log('Decoded barcode:', decoded);
           alert('Decoded barcode: ' + JSON.stringify(decoded));
         } catch (e) {
           console.error('Error decoding barcode', e);
         }
         requestAnimationFrame(scanBarcode);
       };
       requestAnimationFrame(scanBarcode);
     });
   </script>
   ```

In this example, the script:
1. Accesses the device's camera using `navigator.mediaDevices.getUserMedia`.
2. Streams the video feed to a `<video>` element.
3. Captures frames from the video feed and processes them using a `<canvas>`.
4. Attempts to decode the barcode using `pdf417-js`.

This setup runs entirely in the user's browser, ensuring the images are processed locally without needing to upload them to a server.
